File: macos/Sources/WooTodoSync/WebDavSync.swift
import Foundation
import Security

public enum WebDavError: Error, Equatable, LocalizedError, Sendable {
    case invalidCredentials
    case invalidEndpoint
    case encoding(String)
    case transport(String)
    case invalidResponse
    case http(Int)
    case objectConflict(String)
    case malformedObject(String)

    public var errorDescription: String? {
        switch self {
        case .invalidCredentials: "坚果云同步凭据无效"
        case .invalidEndpoint: "坚果云 WebDAV 地址必须是 https://dav.jianguoyun.com/dav/"
        case .encoding(let message): "WebDAV 对象编码失败：\(message)"
        case .transport(let message): "坚果云网络请求失败：\(message)"
        case .invalidResponse: "坚果云返回了无效的 HTTP 响应"
        case .http(let status): "坚果云 WebDAV 返回 HTTP \(status)"
        case .objectConflict(let path): "坚果云对象发生冲突：\(path)"
        case .malformedObject(let message): "坚果云同步对象无效：\(message)"
        }
    }
}

public enum WebDavEndpointPolicy {
    public static let endpoint = URL(string: "https://dav.jianguoyun.com/dav/")!

    public static func isAllowed(_ value: URL) -> Bool {
        guard value.scheme?.lowercased() == "https",
              value.host?.lowercased() == "dav.jianguoyun.com",
              value.user == nil,
              value.password == nil,
              value.query == nil,
              value.fragment == nil else { return false }
        let path = value.path.hasSuffix("/") ? value.path : value.path + "/"
        return path == "/dav/"
    }
}

public struct WebDavCredentials: Codable, Equatable, Sendable {
    public let endpoint: URL
    public let username: String
    public let appPassword: String
    public let vaultId: String
    public let deviceId: String
    public let vaultKey: Data

    public init(
        endpoint: URL = WebDavEndpointPolicy.endpoint,
        username: String,
        appPassword: String,
        vaultId: String,
        deviceId: String,
        vaultKey: Data
    ) {
        self.endpoint = endpoint
        self.username = username
        self.appPassword = appPassword
        self.vaultId = vaultId
        self.deviceId = deviceId
        self.vaultKey = vaultKey
    }

    public func validate() throws {
        guard WebDavEndpointPolicy.isAllowed(endpoint),
              !username.isEmpty,
              !appPassword.isEmpty,
              username.rangeOfCharacter(from: .whitespacesAndNewlines) == nil,
              username.rangeOfCharacter(from: .controlCharacters) == nil,
              appPassword.rangeOfCharacter(from: .controlCharacters) == nil,
              vaultId.range(of: "^[A-Za-z0-9][A-Za-z0-9._-]{0,63}$", options: .regularExpression) != nil,
              deviceId.range(of: "^[A-Za-z0-9][A-Za-z0-9._:-]{7,127}$", options: .regularExpression) != nil,
              vaultKey.count == AES256GCM.keyByteCount else {
            throw WebDavError.invalidCredentials
        }
    }
}

/// 坚果云应用密码与同步密钥只进入 Keychain；WebDAV 服务端永远只看到密文对象。
public final class WebDavCredentialsStore: @unchecked Sendable {
    private let service: String
    private let account: String

    public init(
        service: String = "dev.woo-todo.webdav",
        account: String = "jianguoyun"
    ) {
        self.service = service
        self.account = account
    }

    public func save(_ credentials: WebDavCredentials) throws {
        try credentials.validate()
        let data: Data
        do {
            data = try JSONEncoder().encode(credentials)
        } catch {
            throw WebDavError.encoding(error.localizedDescription)
        }
        var attributes = query()
        attributes[kSecValueData as String] = data
        attributes[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        attributes[kSecAttrSynchronizable as String] = false
        let status = SecItemAdd(attributes as CFDictionary, nil)
        if status == errSecDuplicateItem {
            let updateStatus = SecItemUpdate(
                query() as CFDictionary,
                [kSecValueData as String: data] as CFDictionary
            )
            guard updateStatus == errSecSuccess else {
                throw WebDavError.transport("Keychain 错误码 \(updateStatus)")
            }
        } else if status != errSecSuccess {
            throw WebDavError.transport("Keychain 错误码 \(status)")
        }
    }

    public func load() throws -> WebDavCredentials? {
        var request = query()
        request[kSecReturnData as String] = true
        request[kSecMatchLimit as String] = kSecMatchLimitOne
        var result: CFTypeRef?
        let status = SecItemCopyMatching(request as CFDictionary, &result)
        if status == errSecItemNotFound { return nil }
        guard status == errSecSuccess, let data = result as? Data else {
            throw WebDavError.transport("Keychain 错误码 \(status)")
        }
        do {
            let credentials = try JSONDecoder().decode(WebDavCredentials.self, from: data)
            try credentials.validate()
            return credentials
        } catch let error as WebDavError {
            throw error
        } catch {
            throw WebDavError.invalidCredentials
        }
    }

    public func delete() throws {
        let status = SecItemDelete(query() as CFDictionary)
        guard status == errSecSuccess || status == errSecItemNotFound else {
            throw WebDavError.transport("Keychain 错误码 \(status)")
        }
    }

    private func query() -> [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
    }
}

public struct WebDavOperation: Codable, Equatable, Sendable {
    public let format: String
    public let protocolVersion: Int
    public let vaultId: String
    public let deviceId: String
    public let opId: String
    public let entityId: String
    public let kind: SyncOperationKind
    public let lamport: Int64
    public let nonce: String
    public let ciphertext: String

    public init(
        vaultId: String,
        deviceId: String,
        operation: SyncPushOperation
    ) throws {
        self.format = "woo-todo-webdav-operation"
        self.protocolVersion = 1
        self.vaultId = vaultId
        self.deviceId = deviceId
        self.opId = operation.opId
        self.entityId = operation.entityId
        self.kind = operation.kind
        self.lamport = operation.lamport
        self.nonce = operation.nonce
        self.ciphertext = operation.ciphertext
        try validate()
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: AnyCodingKey.self)
        let actual = Set(container.allKeys.map(\.stringValue))
        let expected = Set([
            "format", "protocolVersion", "vaultId", "deviceId", "opId",
            "entityId", "kind", "lamport", "nonce", "ciphertext",
        ])
        guard actual == expected else {
            throw WebDavError.malformedObject("字段不匹配")
        }
        self.format = try container.decode(String.self, forKey: AnyCodingKey("format"))
        self.protocolVersion = try container.decode(Int.self, forKey: AnyCodingKey("protocolVersion"))
        self.vaultId = try container.decode(String.self, forKey: AnyCodingKey("vaultId"))
        self.deviceId = try container.decode(String.self, forKey: AnyCodingKey("deviceId"))
        self.opId = try container.decode(String.self, forKey: AnyCodingKey("opId"))
        self.entityId = try container.decode(String.self, forKey: AnyCodingKey("entityId"))
        self.kind = try container.decode(SyncOperationKind.self, forKey: AnyCodingKey("kind"))
        self.lamport = try container.decode(Int64.self, forKey: AnyCodingKey("lamport"))
        self.nonce = try container.decode(String.self, forKey: AnyCodingKey("nonce"))
        self.ciphertext = try container.decode(String.self, forKey: AnyCodingKey("ciphertext"))
        try validate()
    }

    public func validate() throws {
        guard format == "woo-todo-webdav-operation",
              protocolVersion == 1,
              vaultId.range(of: "^[A-Za-z0-9][A-Za-z0-9._-]{0,63}$", options: .regularExpression) != nil,
              validIdentifier(deviceId),
              validIdentifier(opId),
              validIdentifier(entityId),
              lamport >= 1,
              (try? Base64URL.decode(nonce).count) == AES256GCM.nonceByteCount,
              (try? Base64URL.decode(ciphertext).count).map({ $0 >= AES256GCM.tagByteCount }) == true else {
            throw WebDavError.malformedObject("元数据或密文长度无效")
        }
    }

    public func pushOperation() -> SyncPushOperation {
        SyncPushOperation(
            opId: opId,
            entityId: entityId,
            kind: kind,
            lamport: lamport,
            ciphertext: ciphertext,
            nonce: nonce
        )
    }

    public static func path(vaultId: String, opId: String) -> [String] {
        let shard = String(opId.prefix(2))
        return ["v1", vaultId, "ops", shard, "\(opId).json"]
    }

    internal static func isValidShard(_ value: String) -> Bool {
        value.range(
            of: "^[A-Za-z0-9][A-Za-z0-9._:-]$",
            options: .regularExpression
        ) != nil
    }

    private func validIdentifier(_ value: String) -> Bool {
        value.range(of: "^[A-Za-z0-9][A-Za-z0-9._:-]{7,127}$", options: .regularExpression) != nil
    }
}

public protocol WebDavLocalApplying: Sendable {
    func applyWebDavOperations(_ operations: [WebDavOperation]) async throws
}

public final class WebDavClient: @unchecked Sendable {
    public let credentials: WebDavCredentials
    private let session: URLSession
    private let baseURL: URL

    public init(credentials: WebDavCredentials, session: URLSession = .shared) throws {
        try credentials.validate()
        self.credentials = credentials
        self.session = session
        self.baseURL = credentials.endpoint
    }

    public func ensureCollections() async throws {
        for path in [
            ["v1"],
            ["v1", credentials.vaultId],
            ["v1", credentials.vaultId, "ops"],
        ] {
            _ = try await request(method: "MKCOL", path: path, body: nil, headers: [:], accepted: [201, 405, 409])
        }
    }

    public func put(_ operation: WebDavOperation) async throws {
        let data: Data
        do {
            let encoder = JSONEncoder()
            encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
            data = try encoder.encode(operation)
        } catch {
            throw WebDavError.encoding(error.localizedDescription)
        }
        let path = WebDavOperation.path(vaultId: credentials.vaultId, opId: operation.opId)
        _ = try await request(
            method: "MKCOL",
            path: Array(path.dropLast()),
            body: nil,
            headers: [:],
            accepted: [201, 405, 409]
        )
        let result = try await request(
            method: "PUT",
            path: path,
            body: data,
            headers: ["Content-Type": "application/json", "If-None-Match": "*"],
            accepted: [200, 201, 204, 405, 409, 412]
        )
        guard [405, 409, 412].contains(result.statusCode) else { return }
        let existing = try await get(path: path)
        guard existing == data else { throw WebDavError.objectConflict(path.joined(separator: "/")) }
    }

    public func listOperationPaths() async throws -> [[String]] {
        let root = ["v1", credentials.vaultId, "ops"]
        let shards = try await propfind(path: root, depth: 1).compactMap { path -> String? in
            guard path.count == root.count + 1,
                  path.dropFirst(root.count).first.map(WebDavOperation.isValidShard) == true
            else { return nil }
            return path.last
        }
        var result: [[String]] = []
        for shard in Set(shards).sorted() {
            let paths = try await propfind(path: root + [shard], depth: 1)
            result.append(contentsOf: paths.filter { path in
                path.count == root.count + 2 && path.last?.hasSuffix(".json") == true
            })
        }
        return result.sorted { $0.joined() < $1.joined() }
    }

    public func get(path: [String]) async throws -> Data {
        try await request(method: "GET", path: path, body: nil, headers: [:], accepted: [200]).data
    }

    private func propfind(path: [String], depth: Int) async throws -> [[String]] {
        let response = try await request(
            method: "PROPFIND",
            path: path,
            body: Data("<?xml version=\"1.0\" encoding=\"utf-8\" ?><propfind xmlns=\"DAV:\"><prop><resourcetype/></prop></propfind>".utf8),
            headers: ["Depth": String(depth), "Content-Type": "application/xml"],
            accepted: [207]
        )
        return try WebDavHrefParser.parse(response.data)
    }

    private func request(
        method: String,
        path: [String],
        body: Data?,
        headers: [String: String],
        accepted: Set<Int>
    ) async throws -> (data: Data, statusCode: Int) {
        var url = baseURL
        for component in path { url.appendPathComponent(component) }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.httpBody = body
        request.timeoutInterval = 20
        request.setValue("Basic \(Data("\(credentials.username):\(credentials.appPassword)".utf8).base64EncodedString())", forHTTPHeaderField: "Authorization")
        for (key, value) in headers { request.setValue(value, forHTTPHeaderField: key) }
        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else { throw WebDavError.invalidResponse }
            guard accepted.contains(http.statusCode) else { throw WebDavError.http(http.statusCode) }
            return (data, http.statusCode)
        } catch let error as WebDavError {
            throw error
        } catch {
            throw WebDavError.transport(error.localizedDescription)
        }
    }
}

public final class WebDavSyncRunner: @unchecked Sendable {
    private let client: WebDavClient
    private let outbox: any SyncOutbox
    private let local: any WebDavLocalApplying

    public init(client: WebDavClient, outbox: any SyncOutbox, local: any WebDavLocalApplying) {
        self.client = client
        self.outbox = outbox
        self.local = local
    }

    public func synchronize() async throws -> SyncRunSummary {
        try await client.ensureCollections()
        let pending = try await outbox.pendingOperations(limit: SyncCoordinator.maximumPushBatch)
        for operation in pending {
            try await client.put(WebDavOperation(
                vaultId: client.credentials.vaultId,
                deviceId: client.credentials.deviceId,
                operation: operation
            ))
        }

        let paths = try await client.listOperationPaths()
        var operations: [WebDavOperation] = []
        for path in paths {
            let data = try await client.get(path: path)
            do {
                let operation = try JSONDecoder().decode(WebDavOperation.self, from: data)
                guard operation.vaultId == client.credentials.vaultId else {
                    throw WebDavError.malformedObject("同步空间不匹配")
                }
                guard path == WebDavOperation.path(
                    vaultId: operation.vaultId,
                    opId: operation.opId
                ) else {
                    throw WebDavError.malformedObject("对象路径与 opId 不匹配")
                }
                operations.append(operation)
            } catch let error as WebDavError {
                throw error
            } catch {
                throw WebDavError.malformedObject(error.localizedDescription)
            }
        }
        for start in stride(from: 0, to: operations.count, by: Self.webDavApplyBatchSize) {
            let end = min(start + Self.webDavApplyBatchSize, operations.count)
            try await local.applyWebDavOperations(Array(operations[start..<end]))
        }
        try await outbox.acknowledgeOperations(opIds: pending.map(\.opId))
        return SyncRunSummary(
            pushed: pending.count,
            pulled: operations.count,
            pages: Self.webDavPageCount(operations.count),
            finalCursor: 0
        )
    }

    internal static let webDavApplyBatchSize = 500

    internal static func webDavPageCount(_ operationCount: Int) -> Int {
        max(1, (operationCount + webDavApplyBatchSize - 1) / webDavApplyBatchSize)
    }
}

private struct AnyCodingKey: CodingKey {
    let stringValue: String
    let intValue: Int? = nil

    init(_ value: String) { stringValue = value }
    init?(stringValue: String) { self.stringValue = stringValue }
    init?(intValue: Int) { stringValue = String(intValue) }
}

internal final class WebDavHrefParser: NSObject, XMLParserDelegate {
    private var current: String?
    private var values: [String] = []

    static func parse(_ data: Data) throws -> [[String]] {
        let parser = WebDavHrefParser()
        let xml = XMLParser(data: data)
        xml.delegate = parser
        guard xml.parse() else { throw WebDavError.malformedObject("PROPFIND XML 无法解析") }
        var paths: [[String]] = []
        for raw in parser.values {
            let value = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !value.isEmpty,
                  let url = URL(string: value),
                  !url.path.isEmpty,
                  let decoded = url.path.removingPercentEncoding else {
                throw WebDavError.malformedObject("PROPFIND href 无效")
            }
            let parts = decoded.split(separator: "/").map(String.init)
            guard let start = parts.firstIndex(of: "v1") else { continue }
            paths.append(Array(parts[start...]))
        }
        return paths
    }

    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String: String] = [:]) {
        if Self.isHrefElement(elementName, qualifiedName: qName) { current = "" }
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        guard let current else { return }
        self.current = current + string
    }

    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        guard Self.isHrefElement(elementName, qualifiedName: qName),
              let current else { return }
        values.append(current)
        self.current = nil
    }

    private static func isHrefElement(_ elementName: String, qualifiedName: String?) -> Bool {
        let candidate = (qualifiedName ?? elementName).split(separator: ":").last.map(String.init)
        return candidate?.caseInsensitiveCompare("href") == .orderedSame
    }
}

File: android/app/src/main/java/com/wootodo/sync/WebDavSync.kt
package com.wootodo.sync

import android.content.Context
import android.net.Uri
import android.util.Xml
import java.io.StringReader
import java.nio.charset.StandardCharsets
import java.util.UUID
import okhttp3.Credentials
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.HttpUrl.Companion.toHttpUrl
import org.json.JSONObject
import org.xmlpull.v1.XmlPullParser

sealed class WebDavException(message: String) : Exception(message) {
    class InvalidCredentials : WebDavException("坚果云同步凭据无效")
    class Transport(message: String) : WebDavException(message)
    class Http(val statusCode: Int) : WebDavException("坚果云 WebDAV 返回 HTTP $statusCode")
    class ObjectConflict(path: String) : WebDavException("坚果云对象发生冲突：$path")
    class MalformedObject(message: String) : WebDavException("坚果云同步对象无效：$message")
}

object WebDavEndpointPolicy {
    const val ENDPOINT = "https://dav.jianguoyun.com/dav/"

    fun isAllowed(value: String): Boolean = runCatching {
        val uri = java.net.URI(value)
        uri.scheme.equals("https", ignoreCase = true) &&
            uri.host.equals("dav.jianguoyun.com", ignoreCase = true) &&
            uri.rawUserInfo == null && uri.rawQuery == null && uri.rawFragment == null &&
            (uri.path == "/dav" || uri.path == "/dav/")
    }.getOrDefault(false)
}

data class WebDavCredentials(
    val endpoint: String = WebDavEndpointPolicy.ENDPOINT,
    val username: String,
    val appPassword: String,
    val vaultId: String,
    val deviceId: String,
    val vaultKey: ByteArray,
) {
    fun validate() {
        if (!WebDavEndpointPolicy.isAllowed(endpoint) ||
            username.isEmpty() || username.any(Char::isWhitespace) || username.any(Char::isISOControl) ||
            appPassword.isEmpty() || appPassword.any(Char::isISOControl) ||
            !VAULT_ID.matches(vaultId) || !IDENTIFIER.matches(deviceId) ||
            vaultKey.size != Aes256Gcm.KEY_BYTES
        ) {
            throw WebDavException.InvalidCredentials()
        }
    }

    fun syncIdentity(): SyncCredentials = SyncCredentials(
        endpoint = endpoint,
        vaultId = vaultId,
        deviceId = deviceId,
        deviceToken = Base64Url.encode(ByteArray(32)),
        vaultKey = vaultKey,
    )

    override fun equals(other: Any?): Boolean = other is WebDavCredentials &&
        endpoint == other.endpoint && username == other.username &&
        appPassword == other.appPassword && vaultId == other.vaultId &&
        deviceId == other.deviceId && vaultKey.contentEquals(other.vaultKey)

    override fun hashCode(): Int = 31 * listOf(
        endpoint,
        username,
        appPassword,
        vaultId,
        deviceId,
    ).hashCode() + vaultKey.contentHashCode()

    private companion object {
        val VAULT_ID = Regex("^[A-Za-z0-9][A-Za-z0-9._-]{0,63}$")
        val IDENTIFIER = Regex("^[A-Za-z0-9][A-Za-z0-9._:-]{7,127}$")
    }
}

class AndroidWebDavCredentialsStore(context: Context) {
    private val blobStore = SharedPreferencesCredentialBlobStore(
        context = context,
        fileName = "webdav_credentials_encrypted",
    )
    private val cipher = AndroidKeystoreCredentialCipher("woo-todo-webdav-credentials-v1")
    private val aad = "woo-todo-webdav-credentials-v1".toByteArray(StandardCharsets.UTF_8)

    @Synchronized
    fun save(credentials: WebDavCredentials) {
        credentials.validate()
        val plaintext = encode(credentials).toByteArray(StandardCharsets.UTF_8)
        try {
            blobStore.write(cipher.encrypt(plaintext, aad))
        } finally {
            plaintext.fill(0)
        }
    }

    @Synchronized
    fun load(): WebDavCredentials? {
        val envelope = blobStore.read() ?: return null
        val plaintext = cipher.decrypt(envelope, aad)
        return try {
            decode(String(plaintext, StandardCharsets.UTF_8)).also(WebDavCredentials::validate)
        } finally {
            plaintext.fill(0)
        }
    }

    @Synchronized
    fun delete() = blobStore.delete()

    private fun encode(credentials: WebDavCredentials): String = JSONObject()
        .put("endpoint", credentials.endpoint)
        .put("username", credentials.username)
        .put("appPassword", credentials.appPassword)
        .put("vaultId", credentials.vaultId)
        .put("deviceId", credentials.deviceId)
        .put("vaultKey", Base64Url.encode(credentials.vaultKey))
        .toString()

    private fun decode(source: String): WebDavCredentials {
        val value = JSONObject(source)
        val expected = setOf("endpoint", "username", "appPassword", "vaultId", "deviceId", "vaultKey")
        require(value.keys().asSequence().toSet() == expected) { "坚果云凭据字段不完整" }
        return WebDavCredentials(
            endpoint = value.getString("endpoint"),
            username = value.getString("username"),
            appPassword = value.getString("appPassword"),
            vaultId = value.getString("vaultId"),
            deviceId = value.getString("deviceId"),
            vaultKey = Base64Url.decode(value.getString("vaultKey")),
        )
    }
}

data class WebDavOperation(
    val vaultId: String,
    val deviceId: String,
    val opId: String,
    val entityId: String,
    val kind: SyncOperationKind,
    val lamport: Long,
    val nonce: String,
    val ciphertext: String,
) {
    fun validate() {
        val identifier = Regex("^[A-Za-z0-9][A-Za-z0-9._:-]{7,127}$")
        if (!Regex("^[A-Za-z0-9][A-Za-z0-9._-]{0,63}$").matches(vaultId) ||
            !identifier.matches(deviceId) || !identifier.matches(opId) ||
            !identifier.matches(entityId) || lamport < 1 ||
            runCatching { Base64Url.decode(nonce).size }.getOrNull() != Aes256Gcm.NONCE_BYTES ||
            (runCatching { Base64Url.decode(ciphertext).size }.getOrNull() ?: 0) < Aes256Gcm.TAG_BYTES
        ) {
            throw WebDavException.MalformedObject("元数据或密文长度无效")
        }
    }

    fun canonicalJson(): String {
        validate()
        return buildString {
            append("{\"ciphertext\":").append(JSONObject.quote(ciphertext))
            append(",\"deviceId\":").append(JSONObject.quote(deviceId))
            append(",\"entityId\":").append(JSONObject.quote(entityId))
            append(",\"format\":\"woo-todo-webdav-operation\"")
            append(",\"kind\":").append(JSONObject.quote(kind.wireValue))
            append(",\"lamport\":").append(lamport)
            append(",\"nonce\":").append(JSONObject.quote(nonce))
            append(",\"opId\":").append(JSONObject.quote(opId))
            append(",\"protocolVersion\":1")
            append(",\"vaultId\":").append(JSONObject.quote(vaultId)).append('}')
        }
    }

    fun metadata(): SyncOperationMetadata = SyncOperationMetadata(
        opId = opId,
        entityId = entityId,
        kind = kind,
        lamport = lamport,
        deviceId = deviceId,
    )

    companion object {
        fun from(push: SyncPushOperation, credentials: WebDavCredentials): WebDavOperation =
            WebDavOperation(
                vaultId = credentials.vaultId,
                deviceId = credentials.deviceId,
                opId = push.opId,
                entityId = push.entityId,
                kind = push.kind,
                lamport = push.lamport,
                nonce = push.nonce,
                ciphertext = push.ciphertext,
            ).also(WebDavOperation::validate)

        fun decode(source: String): WebDavOperation {
            try {
                val value = JSONObject(source)
                val expected = setOf(
                    "format", "protocolVersion", "vaultId", "deviceId", "opId",
                    "entityId", "kind", "lamport", "nonce", "ciphertext",
                )
                fun string(name: String): String = value.opt(name) as? String
                    ?: throw WebDavException.MalformedObject("字段类型不匹配")
                val protocolVersion = value.opt("protocolVersion")
                val lamport = when (val raw = value.opt("lamport")) {
                    is Int -> raw.toLong()
                    is Long -> raw
                    else -> throw WebDavException.MalformedObject("字段类型不匹配")
                }
                if (value.keys().asSequence().toSet() != expected ||
                    string("format") != "woo-todo-webdav-operation" ||
                    !((protocolVersion is Int && protocolVersion == 1) ||
                        (protocolVersion is Long && protocolVersion == 1L))
                ) {
                    throw WebDavException.MalformedObject("字段不匹配")
                }
                return WebDavOperation(
                    vaultId = string("vaultId"),
                    deviceId = string("deviceId"),
                    opId = string("opId"),
                    entityId = string("entityId"),
                    kind = SyncOperationKind.fromWire(string("kind")),
                    lamport = lamport,
                    nonce = string("nonce"),
                    ciphertext = string("ciphertext"),
                ).also(WebDavOperation::validate)
            } catch (error: WebDavException) {
                throw error
            } catch (_: Exception) {
                throw WebDavException.MalformedObject("字段或 JSON 格式无效")
            }
        }

        fun path(vaultId: String, opId: String): List<String> = listOf(
            "v1",
            vaultId,
            "ops",
            opId.take(2),
            "$opId.json",
        )

        internal fun isValidShard(value: String): Boolean =
            Regex("^[A-Za-z0-9][A-Za-z0-9._:-]$").matches(value)
    }
}

interface WebDavLocalApplying {
    fun applyWebDavOperations(operations: List<WebDavOperation>)
}

class WebDavClient(
    val credentials: WebDavCredentials,
    private val httpClient: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(java.time.Duration.ofSeconds(10))
        .readTimeout(java.time.Duration.ofSeconds(20))
        .writeTimeout(java.time.Duration.ofSeconds(20))
        .build(),
) {
    private val baseUrl = credentials.endpoint.toHttpUrl()
    private val authorization = Credentials.basic(credentials.username, credentials.appPassword)

    init {
        credentials.validate()
    }

    fun ensureCollections() {
        listOf(
            listOf("v1"),
            listOf("v1", credentials.vaultId),
            listOf("v1", credentials.vaultId, "ops"),
        ).forEach { request("MKCOL", it, null, emptyMap(), setOf(201, 405, 409)) }
    }

    fun put(operation: WebDavOperation) {
        val data = operation.canonicalJson().toByteArray(StandardCharsets.UTF_8)
        val path = WebDavOperation.path(credentials.vaultId, operation.opId)
        request("MKCOL", path.dropLast(1), null, emptyMap(), setOf(201, 405, 409))
        val result = request(
            method = "PUT",
            path = path,
            body = data,
            headers = mapOf("Content-Type" to "application/json", "If-None-Match" to "*"),
            accepted = setOf(200, 201, 204, 405, 409, 412),
        )
        if (result.code in setOf(405, 409, 412)) {
            val existing = get(path)
            if (!existing.contentEquals(data)) {
                throw WebDavException.ObjectConflict(path.joinToString("/"))
            }
        }
    }

    fun listOperationPaths(): List<List<String>> {
        val root = listOf("v1", credentials.vaultId, "ops")
        val shards = propfind(root)
            .filter {
                it.size == root.size + 1 &&
                    WebDavOperation.isValidShard(it.last())
            }
            .map(List<String>::last)
            .distinct()
            .sorted()
        return shards.flatMap { shard ->
            propfind(root + shard).filter { it.size == root.size + 2 && it.last().endsWith(".json") }
        }.sortedBy { it.joinToString("/") }
    }

    fun get(path: List<String>): ByteArray = request(
        "GET",
        path,
        null,
        emptyMap(),
        setOf(200),
    ).body

    private fun propfind(path: List<String>): List<List<String>> {
        val body = "<?xml version=\"1.0\" encoding=\"utf-8\" ?>" +
            "<propfind xmlns=\"DAV:\"><prop><resourcetype/></prop></propfind>"
        val response = request(
            "PROPFIND",
            path,
            body.toByteArray(StandardCharsets.UTF_8),
            mapOf("Depth" to "1", "Content-Type" to "application/xml"),
            setOf(207),
        )
        return WebDavPropfindParser.parse(String(response.body, StandardCharsets.UTF_8))
    }

    private fun request(
        method: String,
        path: List<String>,
        body: ByteArray?,
        headers: Map<String, String>,
        accepted: Set<Int>,
    ): ResponseData {
        val url = baseUrl.newBuilder().apply {
            path.forEach(::addPathSegment)
        }.build()
        val mediaType = headers["Content-Type"]?.toMediaType()
        val requestBody = if (body == null && method in setOf("PUT", "PROPFIND")) {
            ByteArray(0).toRequestBody(mediaType)
        } else {
            body?.toRequestBody(mediaType)
        }
        val request = Request.Builder()
            .url(url)
            .header("Authorization", authorization)
            .apply { headers.forEach(::header) }
            .method(method, requestBody)
            .build()
        try {
            httpClient.newCall(request).execute().use { response ->
                val responseBody = response.body?.bytes() ?: ByteArray(0)
                if (response.code !in accepted) throw WebDavException.Http(response.code)
                return ResponseData(response.code, responseBody)
            }
        } catch (error: WebDavException) {
            throw error
        } catch (error: Exception) {
            throw WebDavException.Transport(error.localizedMessage ?: "网络不可用")
        }
    }

    private data class ResponseData(val code: Int, val body: ByteArray)
}

class WebDavSyncRunner(
    private val client: WebDavClient,
    private val outbox: OutboxStore,
    private val local: WebDavLocalApplying,
) {
    fun synchronize(): SyncRunSummary {
        client.ensureCollections()
        val pending = outbox.pendingOperations(SyncCoordinator.MAXIMUM_PUSH_BATCH)
        pending.forEach { client.put(WebDavOperation.from(it, client.credentials)) }
        val operations = client.listOperationPaths().map { path ->
            WebDavOperation.decode(String(client.get(path), StandardCharsets.UTF_8)).also { operation ->
                if (operation.vaultId != client.credentials.vaultId) {
                    throw WebDavException.MalformedObject("同步空间不匹配")
                }
                if (path != WebDavOperation.path(operation.vaultId, operation.opId)) {
                    throw WebDavException.MalformedObject("对象路径与 opId 不匹配")
                }
            }
        }
        operations.chunked(WEBDAV_APPLY_BATCH_SIZE).forEach { batch ->
            local.applyWebDavOperations(batch)
        }
        outbox.acknowledgeOperations(pending.map(SyncPushOperation::opId))
        return SyncRunSummary(
            pushed = pending.size,
            pulled = operations.size,
            pages = webDavPageCount(operations.size),
            finalCursor = 0,
        )
    }

    companion object {
        internal const val WEBDAV_APPLY_BATCH_SIZE = 500

        internal fun webDavPageCount(operationCount: Int): Int =
            maxOf(1, (operationCount + WEBDAV_APPLY_BATCH_SIZE - 1) / WEBDAV_APPLY_BATCH_SIZE)
    }
}

internal object WebDavPropfindParser {
    fun parse(source: String): List<List<String>> {
        val parser = Xml.newPullParser().apply {
            // 坚果云常见响应使用 d:href；同时开启命名空间处理并保留本地名兜底。
            setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, true)
            setInput(StringReader(source))
        }
        val paths = mutableListOf<List<String>>()
        var event = parser.eventType
        while (event != XmlPullParser.END_DOCUMENT) {
            if (event == XmlPullParser.START_TAG && isHrefElement(parser.name)) {
                parseHref(parser.nextText())?.let(paths::add)
            }
            event = parser.next()
        }
        return paths
    }

    private fun isHrefElement(name: String?): Boolean =
        name?.substringAfterLast(':')?.equals("href", ignoreCase = true) == true

    private fun parseHref(raw: String): List<String>? {
        val value = raw.trim()
        if (value.isEmpty()) {
            throw WebDavException.MalformedObject("PROPFIND href 无效")
        }
        val rawPath = try {
            java.net.URI(value).rawPath
        } catch (_: Exception) {
            null
        }?.takeIf(String::isNotEmpty)
            ?: throw WebDavException.MalformedObject("PROPFIND href 无效")
        val segments = rawPath
            .split('/')
            .filter(String::isNotEmpty)
            .map(Uri::decode)
        val start = segments.indexOf("v1")
        return start.takeIf { it >= 0 }?.let { segments.drop(it) }
    }
}

fun newWebDavIdentity(): Pair<String, String> =
    "vault-${Base64Url.encode(SecureBytes.generate(9))}" to UUID.randomUUID().toString().lowercase()
